cognite_toolkit/_cdf_tk/resource_classes/container_field_definitions.py

```diff
@@ -0,0 +1,361 @@
+
+
+class UniquenessConstraintDefinition(ConstraintDefinition):
+    _constraint_type = "uniqueness"
+    constraint_type: Literal["uniqueness"] = "uniqueness"
+    properties: list[str] = Field(description="List of properties included in the constraint.")
+    by_space: bool | None = Field(default=None, description="Whether to make the constraint space-specific.")
+
+
+class RequiresConstraintDefinition(ConstraintDefinition):
+    _constraint_type = "requires"
+    constraint_type: Literal["requires"] = "requires"
+    require: ContainerReference = Field(description="Reference to an existing container.")
+
+
+_CONSTRAINT_DEFINITION_CLASS_BY_TYPE: MappingProxyType[str, type[ConstraintDefinition]] = MappingProxyType(
+    {c._constraint_type: c for c in ConstraintDefinition.__subclasses__()}
+)
+
+
+class IndexDefinition(BaseModelResource):
+    _index_type: ClassVar[str]
+    index_type: str
+    properties: list[str] = Field(description="List of properties to define the index across.")
+
+    @model_validator(mode="wrap")
+    @classmethod
+    def find_index_definition_cls(cls, data: Any, handler: ModelWrapValidatorHandler[Self]) -> Self:
+        if isinstance(data, IndexDefinition):
+            return cast(Self, data)
+        if not isinstance(data, dict):
+            raise ValueError(f"Invalid index definition data '{type(data)}' expected dict")
+
+        if cls is not IndexDefinition:
+            return handler(data)
+
+        index_type = data.get("indexType")
+        if index_type is None:
+            raise ValueError("Missing 'indexType' field in index definition data")
+        if index_type not in _INDEX_DEFINITION_CLASS_BY_TYPE:
+            raise ValueError(
+                f"invalid index type '{index_type}'. Expected one of {humanize_collection(_INDEX_DEFINITION_CLASS_BY_TYPE.keys(), bind_word='or')}"
+            )
+        cls_ = _INDEX_DEFINITION_CLASS_BY_TYPE[index_type]
+        return cast(Self, cls_.model_validate(data))
+
+    @model_serializer(mode="wrap", when_used="always", return_type=dict)
+    def serialize_index(self, handler: SerializerFunctionWrapHandler) -> dict:
+        serialized_data = handler(self)
+        return serialized_data
+
+
+class BtreeIndex(IndexDefinition):
+    _index_type = "btree"
+    index_type: Literal["btree"] = "btree"
+    by_space: bool | None = Field(default=None, description="Whether to make the index space-specific.")
+    cursorable: bool | None = Field(
+        default=None, description="Whether the index can be used for cursor-based pagination."
+    )
+
+
+class InvertedIndex(IndexDefinition):
+    _index_type = "inverted"
+    index_type: Literal["inverted"] = "inverted"
+
+
+_INDEX_DEFINITION_CLASS_BY_TYPE: MappingProxyType[str, type[IndexDefinition]] = MappingProxyType(
+    {c._index_type: c for c in IndexDefinition.__subclasses__()}
+)
+
+
+class PropertyTypeDefinition(BaseModelResource):
+    _property_type: ClassVar[str]
+    type: str
```

Comment:
Why do you have both a type classvar and an attribute type? Suggest removing `type` and only keep the `_property_type`. Look at how we do it in the `cognite-sdk`: https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/data_classes/data_modeling/data_types.py#L49